tests/frontier/opcodes/test_callcode.py

```diff
@@ -0,0 +1,117 @@
+"""
+abstract: Tests the CALLCODE gas consumption with a CALL/CALLCODE value transfer.
+
+    Tests an EthereumJS bug where the CALLCODE gas was incorrectly calculated from a
+    CALL value transfer: https://github.com/ethereumjs/ethereumjs-monorepo/issues/3194
+
+    Test setup: 2 smart contract accounts exist where AAAB CALL/CALLCODEs AAAA, which then
+    CALLCODEs a non-existent contract AAAC. The result of AAAA's CALL/CALLCODE is stored in AAAB's
+    storage slot 0.
+"""
+
+from typing import Dict
+
+import pytest
+
+from ethereum_test_tools import (
+    Account,
+    Environment,
+    StateTestFiller,
+    TestAddress,
+    Transaction,
+    to_address,
+)
+from ethereum_test_tools.vm.opcode import Opcodes as Op
+
+
+@pytest.fixture
+def caller_code(caller_gas, caller_type):
+    """
+    Code to call the callee contract:
+        PUSH1 0x00
+        DUP1 * 4
+        PUSH2 0xAAAA
+        PUSH2 call_gas
+        CALL/CALLCODE
+        PUSH1 0x00
+        SSTORE
+    """
+    return Op.SSTORE(0, caller_type(caller_gas, 0xAAAA, 0, 0, 0, 0, 0))
+
+
+@pytest.fixture
+def callee_code():
+    """
+    Code called by the caller contract:
+        PUSH1 0x00 * 4
+        PUSH1 0x00
+        PUSH1 0x00
+        PUSH2 0xAACC
+        PUSH2 0x1a90
+        CALLCODE
+    """
+    return Op.CALLCODE(0x1A90, 0xAACC, 1, 0, 0, 0, 0)
+
+
+@pytest.fixture
+def caller_tx() -> Transaction:
+    """
+    Transaction that calls the callee contract.
+    """
+    return Transaction(
+        chain_id=0x01,
+        nonce=0,
+        to=to_address(0xAAAB),
+        value=1,
+        gas_limit=500000,
+        gas_price=7,
+    )
+
+
+@pytest.fixture
+def pre(callee_code, caller_code) -> Dict[str, Account]:  # noqa: D103
+    return {
+        to_address(0xAAAA): Account(
+            balance=0x03,
+            code=callee_code,
+            nonce=1,
+        ),
+        to_address(0xAAAB): Account(
+            balance=0x03,
+            code=caller_code,
+            nonce=1,
+        ),
+        TestAddress: Account(
+            balance=0x0BA1A9CE,
+        ),
+    }
+
+
+@pytest.fixture
+def post(caller_gas) -> Dict[str, Account]:  # noqa: D103
+    return {
+        to_address(0xAAAB): Account(storage={0x00: (0x01 if caller_gas >= 0x3D50 else 0x00)}),
+    }
+
+
+@pytest.mark.parametrize(
+    "caller_gas",
+    [0x2D5F, 0x3D50],
+    ids=["insufficient", "sufficient"],
+)
+@pytest.mark.parametrize(
+    "caller_type",
+    [Op.CALL, Op.CALLCODE],
+)
+@pytest.mark.valid_from("London")
```

Comment:
These tests are now inside the Frontier folder but they are only valid from London... is this allowed?

Reply:
Yeah, we've been discussing test placement this week! I do think what we have can be a little confusing, and very open to changing it.

The general rule for now: tests should be added to the fork folder where the feature being tested was added. So in this case, as the callcode opcode was added in Frontier, we add the test here, even though the test is only valid from London.
```
mark_valid_from_fork >= feature_introduced_fork
```
This is to sync more with the structure of eip tests, i.e the 4844 eip tests are in the Cancun folder, as 4844 is introduced in Cancun.

We will add formal rules to the docs and checks to the framework for this (soon-ish).

Reply:
Ideally in the future we will have a lot more callcode tests. So the primary usecase for this structure is to keep all the callcode tests in one place, as opposed to multiple different folders.

Will think on this some more!

Reply:
Gave this some more thought but this actually makes a lot of sense. For instance, if we would now move these tests into London, then later on if we want to add a specific Chainstart/Frontier test then we would have to `git mv` it into that hardfork folder. So I agree, put the tests in the hardfork folder where the opcode has been introduced. And of course, changes later (for instance gas cost of callcode) could be marked specific in a certain hardfork range if those are added in the same file.

Reply:
I'm glad you agree :) and in that case we shall stick to this! I'm hoping to add this information to future docs very soon, alongside checks within the framework for the: `mark_valid_from_fork` pytest directive.